### PR TITLE
Move to choice monad

### LINKED ISSUE #20: Monad Result
Change the code to be functional and use the Result monad to manage error.

### PR DESCRIPTION
**Pull request by maiste**
Move all the modules to the choice monad and update tests.
Fixes #20

### CODE REVIEW

File: tests/test_conf.ml [hunk 1]
```diff
@@ -4,35 +4,43 @@
...
-    (check (option (list string))) ("regex_content : "^ a) b (apply_option (List.sort compare) (Conf.get_file_regex js))
+    let js = match Conf.init_json ("res/" ^ a) with
+      | Ok elt -> elt
+      | _ -> raise Test_error 
+    in 
+    let res = match Conf.get_file_regex js with
+      | Ok content -> (List.sort compare content)
+      | Error elt -> []
+    in
+    (check (bool)) ("regex_content : "^ a) true (b=res)
```
Comment (xavitator) [inline tests/test_conf.ml]: replace `(check (bool)) ("regex_content : "^ a) true (b=res)` by `(check (list string)) ("regex_content : "^ a) b res` : compare string list with the right comparator.

File: tests/test_finder.ml [hunk 2]
```diff
@@ -44,17 +44,23 @@ let finder_cmd rgxl =
...
   let aux turn h =
     let res_cmd = List.sort compare (finder_cmd h) in
-    let res_finder = List.sort compare ("" :: Finder.get_all_files h) in
-    (check (list string)) ("equal list " ^ (string_of_int turn)) res_cmd res_finder
+    let res_finder = match Finder.get_files h with 
+      | Ok content -> List.sort compare (""::content)
+      | Error str -> [""]
+    in 
+    (check (list string)) 
+      ("accequal list " ^ (string_of_int turn)) res_cmd res_finder
```
Comment (xavitator) [inline tests/test_finder.ml]: I do not understand what 'accequal' means ?

File: tests/test_finder.ml [hunk 3]
```diff
@@ -44,17 +44,23 @@ let finder_cmd rgxl =
...
-
 (* Test if finder works correctly *)
 let test_finder_eq _ = 
   let aux turn h =
     let res_cmd = List.sort compare (finder_cmd h) in
-    let res_finder = List.sort compare ("" :: Finder.get_all_files h) in
-    (check (list string)) ("equal list " ^ (string_of_int turn)) res_cmd res_finder
+    let res_finder = match Finder.get_files h with 
+      | Ok content -> List.sort compare (""::content)
+      | Error str -> [""]
```
Comment (xavitator) [inline tests/test_finder.ml]: It will, maybe, change with the comment in `finder.ml`.

**Review by xavitator — CHANGES_REQUESTED** [review]
A general comment for `test_reader.ml` : we could add more test about empty file, file with some empty line between two filled lines, etc.

File: tests/test_finder.ml (hunk 2 above)
Comment (maiste) [reply tests/test_finder.ml]: Typo.

File: tests/test_conf.ml (hunk 1 above)
Comment (maiste) [reply tests/test_conf.ml]: It's your test. I don't change anything from the previous test.

File: src/core/reader.ml [hunk 4]
```diff
@@ -4,39 +4,42 @@
...
+  | Failed of string
 
 (* Create an input channel for [filename] file *)
-let open_file filename : in_channel option = 
+let open_file filename : in_channel choice = 
   try 
-    Some (open_in filename)
-  with Sys_error _ -> None
+    Ok (open_in filename)
+  with Sys_error _ -> Error "[Error] Can't open file"
```
[the remark this answers is not in the document]
Comment (maiste) [reply src/core/reader.ml]: I think it's an improvement : not for the basic system. I want to improve the error management system in another part :wink:

File: src/core/reader.mli [hunk 5]
```diff
@@ -3,5 +3,9 @@
  * DURAND - MARAIS © 2019
  *)
 
-val read_file : string -> string list option 
+open Monad
+
+(** Return the content of the file named by [str] and return an either
+ * type *)
```
[the remark this answers is not in the document]
Comment (maiste) [reply src/core/reader.mli]: Typo.

File: src/core/conf.ml [hunk 6]
```diff
@@ -5,80 +5,75 @@
...
-let string_of_json = function
-  | `Assoc _ -> "assoc"
-  | `Bool _ -> "boolean"
-  | `Float _ -> "float"
-  | `Int _ -> "integer"
-  | `List _ -> "List"
-  | `Null -> "null"
-  | `String _ -> "string"
+  try Ok (from_file path)
+  with Sys_error _ -> Error "[Error] can't init json"
```
[the remark this answers is not in the document]
Comment (maiste) [reply src/core/conf.ml]: ?

File: src/core/conf.ml [hunk 7]
```diff
@@ -5,80 +5,75 @@
...
+
 
 (* Get the corresponding fiel of [str] in [json] *)
 let member str js = 
   try
-    apply_option (Util.member str) js
-  with Util. Type_error (msg, _) -> let _ = msg in (* print error msg *) None
+    Ok (Util.member str js)
+  with Util. Type_error _ -> Error "[Error] can't get json field"
```
[the remark this answers is not in the document]
Comment (maiste) [reply src/core/conf.ml]: Error message will be improved in another part.

File: src/core/finder.ml [hunk 8]
```diff
@@ -69,3 +71,9 @@ let get_all_files target : string list =
     Array.fold_left aux acc all_f
   in
   get_file target "." []
+
+
+let get_files target : string list choice = 
+  match get_all_files target with
+  | [] -> Error "[Error] can't find any file"
```
[the remark this answers is not in the document]
Comment (maiste) [reply src/core/finder.ml]: Not tail recursive. 
Can we say it's another if there isn't any targeted file ?

[the remark this answers is not in the document]
Comment (maiste) [reply src/core/finder.ml]: Change with [warning]

**Review by xavitator — CHANGES_REQUESTED** [review]

**Review by xavitator — APPROVED** [review]
